agent/agent.go

```diff
@@ -0,0 +1,431 @@
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsRemoteConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsEffectiveConfig |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsOwnMetrics |
+			protobufs.AgentCapabilities_AgentCapabilities_AcceptsOpAMPConnectionSettings |
+			protobufs.AgentCapabilities_AgentCapabilities_ReportsHealth,
+	}
+
+	err := agent.opampClient.SetAgentDescription(agent.agentDescription)
+	if err != nil {
+		return err
+	}
+	err = agent.opampClient.SetHealth(healthMessage(false))
+	if err != nil {
+		return err
+	}
+
+	agent.opampClient.SetCustomCapabilities(&protobufs.CustomCapabilities{
+		Capabilities: []string{sendAgentTelemetryCapability},
+	})
+
+	agent.logger.Debugf(context.Background(), "starting opamp client")
+
+	err = agent.opampClient.Start(context.Background(), settings)
+	if err != nil {
+		return err
+	}
+	agent.logger.Debugf(context.Background(), "started opamp client")
+
+	go agent.healthCheck()
+	go agent.reportUsagePeriodically()
+	return nil
+}
+
+func (agent *Agent) Stop(ctx context.Context) {
+	agent.logger.Debugf(ctx, "disconnecting from OpAMP server")
+	err := agent.opampClient.SetHealth(
+		&protobufs.ComponentHealth{
+			Healthy: false, LastError: "Refinery is shutdown",
+		},
+	)
+	if err != nil {
+		agent.logger.Errorf(ctx, "Could not report health to OpAMP server: %v", err)
+	}
+	err = agent.opampClient.Stop(ctx)
+	if err != nil {
+		agent.logger.Errorf(ctx, "Failed to stop OpAMP client: %v", err)
+	}
+
+	agent.cancel()
+}
+
+func (agent *Agent) healthCheck() {
+	//TODO: make this ticker configurable
+	timer := agent.clock.NewTicker(15 * time.Second)
+	for {
+		select {
+		case <-agent.ctx.Done():
+		case <-timer.Chan():
+			report := agent.calculateHealth()
+			if report != nil {
+				agent.lastHealth = report
+				if err := agent.opampClient.SetHealth(report); err != nil {
+					agent.logger.Errorf(context.Background(), "Could not report health to OpAMP server: %v", err)
+				}
+			}
+
+			traceUsage, ok := agent.metrics.Get("bytes_received_traces")
+			if !ok {
+				agent.logger.Errorf(context.Background(), "unexpected missing trace usage metric")
+			}
+			logUsage, ok := agent.metrics.Get("bytes_received_logs")
+			if !ok {
+				agent.logger.Errorf(context.Background(), "unexpected missing log usage metric")
+			}
+
+			agent.usageTracker.Add(signal_traces, traceUsage)
+			agent.usageTracker.Add(signal_logs, logUsage)
+		}
+	}
+}
+
+func (agent *Agent) reportUsagePeriodically() {
+	timer := agent.clock.NewTicker(15 * time.Second)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-agent.ctx.Done():
+			// TODO: drain the existing reports
+			return
+		case <-timer.Chan():
+			if err := agent.sendUsageReport(); err != nil {
+				if errors.Is(err, errNoData) {
+					agent.logger.Debugf(context.Background(), "No data to report")
+					continue
+				}
+				agent.logger.Errorf(context.Background(), "MONEY STEALING. Could not send usage report: %v", err)
```

Comment:
Should we change this error message to something more user friendly?

Reply:
Fixed 😄